traits/observation/tests/test_observe.py

```diff
@@ -653,3 +655,72 @@ def test_apply_observers_different_target(self):
         # then
         # the handler should be called twice as the targets are different.
         self.assertEqual(handler.call_count, 2)
+
+
+# ---- Low-level tests for async dispatch_same ------------------------------
+
+
+class TestAsyncDispatchSame(unittest.IsolatedAsyncioTestCase):
+    """Test low-level async dispatch."""
+
+    def setUp(self):
+        from traits.observation.observe import _active_handler_tasks
+
+        # ensure no lingering references to handler tasks after test run
+        self.addCleanup(_active_handler_tasks.clear)
+
+        push_exception_handler(reraise_exceptions=True)
+        self.addCleanup(pop_exception_handler)
+
+    async def test_async_dispatch(self):
+        event = asyncio.Event()
+
+        async def handler(event):
+            event.set()
+
+        dispatch_same(handler, event)
+
+        await asyncio.wait_for(event.wait(), timeout=10)
+
+        self.assertTrue(event.is_set())
+
+    async def test_async_dispatch_error(self):
+        event = asyncio.Event()
+        exceptions = []
+
+        async def handler(event):
+            raise Exception("Bad handler")
+            event.set()
+
+        def exception_handler(loop, context):
+            exceptions.append(context["exception"].args[0])
+
+        with self.asyncio_exception_handler(exception_handler):
+            with self.assertRaises(asyncio.exceptions.TimeoutError):
+                dispatch_same(handler, event)
+                await asyncio.wait_for(event.wait(), timeout=0.1)
```

Comment:
I adjusted this slightly in cc9127caa6588095f48162afd61f9c4523b72dca to move the event to the exception handler, which then gives us something to wait on.